emac HAL API, WiFiInterface additions

0xc0170 opened the pull request:
## Description

This PR contains:
- wifi extensions - connect with extended parameter lists, connect async, scan and scan asyn
- emac API
- ublox emac

Notes for reviewers:
@geky raised a question about those async methods, that if we should not be consistent and provide `attach`, `set_blocking` instead of `connect_async` for instance. I haven't yet reworked this as it will impact the esp8266 update and ublox driver changes, thus once we are all good with the rest, we can consider this, and I can redesign it (while there are still as pull request and we can edit those).

The branch was shared for the development, therefore there are merge commits (updates from master to bring network changes like lwip2 and others), and some commits are not that clean (additions and removals later). I can clean it up as the last step. Please do not focus on git log at first.

I can also split ublox emac HAL implementation from this patch.

What's missing are tests, they will come separately, plus ublox implementation (+driver).
## Status

**IN DEVELOPMENT** 
Up for review now, still expecting some changes. Need to run some basic tests as well, ethernet should be retested that any changes in lwip_stack files do not affect its functionality.
## Migrations

The wifi modules might need to extend their implementation, for instance, there's esp8266 pull request done by @bulislaw  : https://github.com/ARMmbed/esp8266-driver/pull/11/files
## Todos
- [ ] Tests
- [x] Documentation (API at least contains doxygen)

please review @kjbracey-arm @c1728p9 @geky @bulislaw @sg- @andreaslarssonublox @lauri-piikivi

If anyone can help to test basic eth functionality, would be appreciated.

bulislaw committed:
Extend and rework WiFi interface

General refactoring of the API and new methods added:
  * get_rssi() - measures radio signal strenght
  * get_state() - returns current state (not connected,
   		 connecting, connected, error)
  * scan() - scans for available networks sync and async versions
  * connect_async() - connect to a network without blocking

0xc0170 committed:
Merge pull request #2699 from bulislaw/feature_wifi_rebase

Feature wifi rebase

0xc0170 committed:
Merge pull request #2627 from bulislaw/feature_wifi

Extend and rework WiFi interface

bulislaw committed:
HAL: Add Emac interface

bulislaw committed:
EMAC: Add emac_stack_t implementation for LWIP

bulislaw committed:
EMAC: Add LWIP adaptation layer

bulislaw committed:
EMAC: Add LWIP adaptation layer

bulislaw committed:
WiFi: Add IPStackInterface abstraction for IP stacks

bulislaw committed:
WiFi: Add IPStackInterface abstraction for IP stacks

bulislaw committed:
WiFi: Extend WiFiInterface API and documentation

* Add details to documentation
* Add timeout for async functions
* Add channel parameter to connect
* Add wifi_ap_t to connection cb function

Commit:
Fixed return type for disconnect

Commit:
Added EMAC

Commit:
Added pointer for wifi_ap_t in wifi_connect_cb_t to be able to use NULL

Commit:
Added missing ;

Commit:
Added wifi_emac_api.cpp

0xc0170 committed:
Merge pull request #2729 from bulislaw/feature_wifi

HAL: Add Emac interface

bulislaw committed:
EMAC: Fix "#if DEVICE_EMAC " usage

0xc0170 committed:
Merge branch 'master' into feature_wifi

Conflicts:
	features/net/FEATURE_IPV4/lwip-interface/EthernetInterface.cpp
	features/net/FEATURE_IPV4/lwip-interface/lwip_stack.c
	features/net/FEATURE_IPV4/lwip-interface/lwip_stack.h
	features/net/network-socket/WiFiInterface.h

bulislaw committed:
WiFi: Decuple IP stack bringup and IP address provisioning

Commit:
Removed check since the interface is not fully up yet so we can't block DHCP/ARP

0xc0170 committed:
K64F: add EMAC device_has

0xc0170 committed:
lwip: add emac interrupt API

ethernet uses eth_arch_xxx_interrupts, emac_xxx_enable_interrupts

Commit:
Added start_dhcp and start_static_ip

Commit:
Fixed missing timeout argument for lwip_start_dhcp

Commit:
Fixed bug in packetIndication for packetInfo->rxData + use mutexes + cleanup

Commit:
Added

0xc0170 committed:
lwipipstack: lwip_bringup arguments update

It now accepts additional arguments, that lwipipstack do not use, thus
all 0 and false for dhcp.

0xc0170 committed:
network-sockets: revert API for static configuration

WifiInterface - add set_credentials

0xc0170 committed:
WiFiInterface: remove connect implementation

We do not want to mix Ethernet and Wifi at the moment, thus WiFiInterface
should implement own connect using emac.

0xc0170 committed:
Merge branch 'master' into feature_wifi

Conflicts:
	features/net/FEATURE_IPV4/lwip-interface/lwip_stack.c
	hal/targets.json

0xc0170 committed:
lwip emac - remove non-existent header inclusion

0xc0170 committed:
IPStackInterface - add ip address

Currently only for IPv4. lwip was updated to accept 2 arguments for getting
the ip address (buffer and length).

0xc0170 committed:
WifiInterface - use Callback class for callbacks

0xc0170 committed:
WiFiInterface: change scope of wifi_ap_t structure

0xc0170 committed:
nsapi: error no ssid name correction

0xc0170 committed:
EthernetInterface: ctor uses connect() method

0xc0170 committed:
lwipstack: add start dhcp and static ip functions

0xc0170 committed:
eth arch: missing endif fix

0xc0170 committed:
eth arch: duplication of declarations removal

0xc0170 committed:
IPStackInterface: removal of this layer

Use NetworkInterface that defines the same API (extended to cover some additional
requirements that comes from wifi).

0xc0170 committed:
WiFiInterface: change variables to protected

A wifi instance should be able to set those data.

0xc0170 committed:
targets: remove emac for k64f, not supported yet

0xc0170 committed:
lwip stack: dhcp in bringup not done for emac targets

0xc0170 committed:
EthernetInterface: revert unrelated changes

0xc0170 committed:
lwip_stack: remove emac int enable/disable functions

Already done in emac interface.

0xc0170 committed:
NetworkInterface- fix comments for dhcp and ip methods

0xc0170 committed:
NetworkInterface: fix start_dhcp argument type

0xc0170 committed:
WiFiInterface - remove data arguments

Callback class provides API to support class callbacks.

0xc0170 committed:
NetworkInterface: remove dhcp and static ip methods

These use cases are already supported by NetworkInterface class via
set_dhcp, and set_network.

0xc0170 committed:
WiFiInterface - revert removal of connect method

0xc0170 committed:
emac interface - typedef should be available for non-emac targets

As it's required by lwip_bringup function

geky committed:
wifi - Removed asynchronous functions

Current asynchronous functions do not match the blocking/attach
pattern of the current codebase.

Unfortunately, the asynchronous api can not be updated in the
current time constraints. The async functions will be removed
for now and can be reconsidered at a later date.

geky committed:
wifi - Moved wifi_ap_t into its own class WiFiAccessPoint

Using a class for this structure is more idiomatic C++, allows
future deprecations, and more flexibility of the internal structure.

Additionally, this matches to design of the SocketAddress class
specific to the network-socket api.

geky committed:
cellular - Matched changes to the WiFiinterface in cellular

0xc0170 committed:
UBLOX_C029: remove emac implementation

It depends on the driver that will come separately, therefore this will be
enabled once driver is in place.

geky committed:
Merge commit 'upstream/master~' into feature_wifi_ublox

geky committed:
Merge remote-tracking branch 'upstream/master' into feature_wifi_ublox_merge

Commit:
Added function for increasing ref count.

kjbracey reviewed (commented):
Couple of style notes. Bit sad that no IPv6 support in emac_lwip.c.

Not supporting IPv6 yourself is one thing, but I think it the current emac_lwip.c will break compilation if IPv4 is disabled in the build.

bulislaw committed:
EMAC: Fix LWIP_IPV4 and DEVICE_EMAC flag usage

Commit:
Fixed missing abstract connect method

Commit:
Fixed missing public

Commit:
Increased the stack size for the lwIP thread from 1024 to 1200. Needed for the u-blox ODIN-W2 driver. The actually needed size has not been measured.

bulislaw committed:
Refactor lwip_stack function names to include mbed prefix

Mbed function names, prefixed with lwip are confusing, as the name
suggest that they are part of lwip ip stack.

bulislaw committed:
Split lwip initialisation and interface bringup

Split IP stack initialisation from mbed_lwip_bringup to mbed_lwip_init.

bulislaw committed:
Fix merge issues

Commit:
Fixed mbed_lwip_get_netmask amd mbed_lwip_get_gateway

0xc0170 committed:
Merge pull request #2 from andreaslarssonublox/feature_wifi_ublox_andreas

wifi: ublox fixes to lwip and wifi

bulislaw committed:
Fix spelling error: NSAPI_SECURITY_UNSUPPORTED

geky committed:
Renamed NSAPI_SECURITY_UNSUPPORTED -> NSAPI_SECURITY_UNKNOWN

Returning a wifi access point without information regarding the
security type is only valid if the security type is unknown (from
the perspective of the network-socket API). For clarity in situations
in which scan may return an unsupported, but known security type,
type name has been changed to NSAPI_SECURITY_UNKNOWN.
